batch/src/main/java/org/openmrs/analytics/JdbcFhirMode.java

```diff
@@ -0,0 +1,233 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package org.openmrs.analytics;
+
+import java.beans.PropertyVetoException;
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import ca.uhn.fhir.context.FhirContext;
+import com.google.gson.Gson;
+import com.mchange.v2.c3p0.ComboPooledDataSource;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.io.FileIO;
+import org.apache.beam.sdk.io.jdbc.JdbcIO;
+import org.apache.beam.sdk.io.parquet.ParquetIO;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.GroupByKey;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.hl7.fhir.dstu3.model.Resource;
+import org.openmrs.analytics.model.EventConfiguration;
+import org.openmrs.analytics.model.GeneralConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JdbcFhirMode {
+	
+	private static final Logger log = LoggerFactory.getLogger(JdbcFhirMode.class);
+	
+	JdbcFhirMode() {
+	}
+	
+	public JdbcIO.DataSourceConfiguration getJdbcConfig(FhirEtl.FhirEtlOptions options) throws PropertyVetoException {
+		ComboPooledDataSource dataSource = new ComboPooledDataSource();
+		dataSource.setDriverClass(options.getJdbcDriverClass());
+		dataSource.setJdbcUrl(options.getJdbcUrl());
+		dataSource.setUser(options.getDbUser());
+		dataSource.setPassword(options.getDbPassword());
+		dataSource.setMaxPoolSize(options.getJdbcMaxPoolSize());
+		dataSource.setInitialPoolSize(10); // TODO: make this configurable if issues arises
+		return JdbcIO.DataSourceConfiguration.create(dataSource);
+	}
+	
+	public static PCollection<KV<String, Iterable<Integer>>> createChunkRanges(String tableName, int fetchSize, Pipeline p,
+	        JdbcIO.DataSourceConfiguration config) {
+		String tableId = tableName + "_id";
+		return p.apply(String.format("Read Max table Index Column for: %s", tableName),
+		    JdbcIO.<String> read().withDataSourceConfiguration(config)
```

Comment:
Any thoughts about this suggestion? It seems to me that you are still getting the max and creating the segments through Beam which is okay but seems unnecessary if we can get similar performance by doing these steps outside Beam.

Reply:
Thanks @kimaina for creating issue #88 for this but I prefer if we resolve this in this PR. The reason is that if my reasoning is correct, then we are adding some non-trivial amount of complexity here (the two PTransforms for getting the MAXs and creating chunk creation) and then leaving it for future to simplify. I feel the right approach is to actually start simple. I understand your Beam logic but I feel it is unnecessarily complex.

It seems to me that you already have the whole logic, you just need to take it out of PTransform and create simple functions for it. That way, it is easier and more trivial to unit-test, e.g., your segmentation logic (i.e., chunk creation).

Of course, all of this is with the assumption that these steps are quick, so if you tell me that is not the case and in fact Beam parallelization helps a lot (e.g., for segmentation), then that is a different story, but I really doubt that is the case.

Reply:
Thanks, @bashir2 for your quick feedback! The reason why we would rather have this implementation on a separate ticket is because of the TODO in line 104. Why? because the logic for `FetchMaxId` will change once we address #46  Does this makes sense? Happy to explain further.

Reply:
> That way, it is easier and more trivial to unit-test, e.g., your segmentation logic (i.e., chunk creation).

It is actually tractable to test any PTransform, please see how this has been done in `batch/src/test/java/org/openmrs/analytics/JdbcFhirModeTest.java`

Reply:
To sum up: Are we expecting the implementation of the join logic proposed in #46 to require Beam parallelization/distribution? If the answer is no then let's switch the first two steps to non-Beam now to make the pipeline code simpler.

Reply:
> I have written many DoFns/PTransforms and am okay understanding the current logic, but that is a general guidelines that I have found very useful over the years, i.e., to avoid Beam constructs if parallelization/distribution is not needed.

Thanks @bashir2  I think this is the missing piece that was not clear to me. My understanding was the other way round, i.e try as much as possible to model your pipeline inside the runner. Still new to beam! I Will definitely move the logic on range generation outside beam